submissions/Oleksandr-Heleta/task_js_oop/app.js

```diff
@@ -0,0 +1,131 @@
+const START_X = 202,
+    START_Y = 404,
+    CELL_WIDTH = 100,
+    CELL_HEIGTH = 83,
+    ROW_POSITION = [60, 143, 228],
+    FIELD_WIDTH = CELL_WIDTH * 5,
+    ENEMY_SPEED = {
+        min: 60,
+        max: 200,
+    };
+
+
+const counter = document.createElement("div");
+let count = 0;
+counter.innerHTML = `Counter: ${count}`;
+document.body.append(counter);
+
+const Person = function (x, y) {
+    this.x = x;
+    this.y = y;
+};
+
+Person.prototype.render = function () {
+    ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
+};
+
+var Enemy = function (x, y, speed, player) {
+    this.x = x;
+    this.y = y;
+    this.speed = this.getRandomSpeed(speed);
+    this.player = player
+    this.sprite = 'images/enemy-bug.png';
+};
+
+const Player = function (x, y) {
+    this.x = x;
+    this.y = y;
+    this.sprite = "images/char-boy.png";
+};
+
+Enemy.prototype = Object.create(Person.prototype);
+Player.prototype = Object.create(Person.prototype);
+
+Enemy.prototype.update = function (dt) {
+    if (this.x < FIELD_WIDTH) {
+        this.x += this.speed * dt;
+    } else {
+        this.x -= FIELD_WIDTH;
+    }
+    if (this.checkCollision(this.player)) this.player.resetPosition("fail");
+};
+
+Enemy.prototype.checkCollision = function (player) {
+    return (
+        this.y + CELL_HEIGTH > player.y &&
+        player.x < this.x + CELL_WIDTH &&
+        player.x > this.x - CELL_WIDTH
+    );
+};
+
+Enemy.prototype.getRandomSpeed = function ({ min, max }) {
+    let randomSpeed = min + Math.random() * (max + 1 - min);
+    return Math.floor(randomSpeed);
+};
+
+Player.prototype.update = function () {
+    if (this.y > START_Y) {
+        this.y = START_Y;
+    }
+    if (this.y < 0) {
+        this.resetPosition("win");
+    }
+    if (this.x > FIELD_WIDTH - CELL_WIDTH) {
+        this.x -= FIELD_WIDTH;
+    }
+    if (this.x < 0) {
+        this.x = FIELD_WIDTH - CELL_WIDTH;
+    }
+};
+
+Player.prototype.handleInput = function (key) {
+    switch (key) {
+        case "up":
+            this.y -= CELL_HEIGTH;
+            break;
+        case "down":
+            this.y += CELL_HEIGTH;
+            break;
+        case "left":
+            this.x -= CELL_WIDTH;
+            break;
+        case "right":
+            this.x += CELL_WIDTH;
+            break;
+
+        default:
+            break;
+    }
+};
+
+Player.prototype.resetPosition = function (status) {
+    switch (status) {
+        case "win":
+            alert("You win!!!");
+            count += 1;
+            counter.innerHTML = `Counter: ${count}`;
```

Comment:
A well designed class shouldn't know anything about outer environment. What if we want to implement this game on NodeJS? Change of environment shouldn't affect classes. They also shouldn't interact with game environment directly.